domain/relation/state/relation.go

```diff
@@ -1072,33 +1093,187 @@ WHERE  name = $getUnit.name
+	ctx context.Context,
+	relationUUID corerelation.UUID,
+	principalUnitName unit.Name,
+) (*application.ID, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var result *application.ID
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		// Check the relation exists.
+		if relationFound, err := st.checkExistsByUUID(ctx, tx, "relation", relationUUID.String()); err != nil {
 			return errors.Capture(err)
+		} else if !relationFound {
+			return relationerrors.RelationNotFound
+		}
+
+		// Check the relation is alive.
+		if alive, err := st.checkLife(ctx, tx, "relation", relationUUID.String(), life.IsAlive); err != nil {
+			return errors.Errorf("getting relation life: %w", err)
+		} else if !alive {
+			return relationerrors.RelationNotAlive
 		}
 
-		// Fetch all the new settings in the relation for this unit.
-		newSettings, err := st.getRelationUnitSettings(ctx, tx, relationUnitUUID)
+		// Find the unit in the relation.
+		relUnitUUID, unitUUID, err := st.getRelationUnit(ctx, tx, relationUUID, principalUnitName)
 		if err != nil {
-			return errors.Errorf("getting new relation unit settings: %w", err)
+			return errors.Errorf("getting relation unit: %w", err)
+		}
+
+		// Check the unit is alive.
+		if alive, err := st.checkLife(ctx, tx, "unit", unitUUID.String(), life.IsAlive); err != nil {
+			return errors.Errorf("getting unit life: %w", err)
+		} else if !alive {
+			return relationerrors.UnitNotAlive
 		}
 
-		// Hash the new settings.
-		hash, err := hashSettings(newSettings)
+		// Check that we are in a container scoped relation.
+		scope, err := st.getRelationScope(ctx, tx, relationUUID)
 		if err != nil {
-			return errors.Errorf("generating hash of relation unit settings: %w", err)
+			return errors.Errorf("getting relation scope: %w", err)
+		} else if scope != charm.ScopeContainer {
+			return nil
 		}
 
-		// Update the hash in the database.
-		err = st.updateUnitSettingsHash(ctx, tx, relationUnitUUID, hash)
+		// Get the ID of the related subordinate application, if it exists.
+		subAppID, relatedSubExists, err := st.findRelatedSubordinateApplication(ctx, tx, relUnitUUID)
 		if err != nil {
-			return errors.Errorf("updating relation unit settings hash: %w", err)
+			return errors.Errorf("getting related subordinate application: %w", err)
+		} else if !relatedSubExists {
+			return nil
 		}
 
+		// Check if there is already a subordinate unit.
+		if exists, err := st.subordinateUnitExists(ctx, tx, subAppID, unitUUID); err != nil {
+			return errors.Errorf("checking if subordinate already exists: %w", err)
+		} else if exists {
+			return nil
+		}
+
+		result = &subAppID
 		return nil
 	})
-	return errors.Capture(err)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return result, nil
+}
+
+// findRelatedSubordinateApplication returns the application ID of the related
+// subordinate application, and a boolean to indicate if it exists.
```

Comment:
todo: update me. If what exists? The bool returned is if the found application is a subordinate.

Reply:
Clarified.